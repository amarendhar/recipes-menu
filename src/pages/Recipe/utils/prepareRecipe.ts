import { EntryFields, EntrySkeletonType } from "contentful";
import { Maybe, RecipeEntry, RecipeData } from "types";

type RecipeFields = {
  title: EntryFields.Text;
  photo: EntryFields.AssetLink;
  tags: EntryFields.Array<EntryFields.EntryLink<EntrySkeletonType>>;
  description: EntryFields.Text;
  chef: EntryFields.EntryLink<EntrySkeletonType>;
};

export const prepareRecipe = (data: Maybe<RecipeEntry>): Maybe<RecipeData> => {
  if (!data) {
    return null;
  }

  const fields = data?.fields as RecipeFields;

  const id: string = data.sys.id;
  const title = fields?.title || "";
  const image = (fields?.photo?.fields?.file?.url || "") as string;
  const tags =
    fields?.tags?.map((tag) => (tag?.fields?.name as string) || "") || [];
  const description = fields?.description || "";
  const chef = (fields?.chef?.fields?.name || "") as string;
  const rating = 0;

  const recipeData: RecipeData = {
    id,
    title,
    image,
    tags,
    description,
    chef,
    rating,
  };

  return recipeData;
};
