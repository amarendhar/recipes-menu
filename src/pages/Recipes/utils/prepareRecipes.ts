import { Entry, EntryFields, EntrySkeletonType } from "contentful";
import { Maybe, RecipesEntry, RecipesItemData } from "types";

type RecipeFields = {
  title: EntryFields.Text;
  photo: EntryFields.AssetLink;
  tags: EntryFields.Array<EntryFields.EntryLink<EntrySkeletonType>>;
  description: EntryFields.Text;
  chef: EntryFields.EntryLink<EntrySkeletonType>;
};

export const prepareRecipes = (
  data: Maybe<RecipesEntry>
): RecipesItemData[] => {
  if (!data?.items) {
    return [];
  }

  const recipes: RecipesItemData[] = data.items.map(
    (recipe: Entry<EntrySkeletonType, undefined, string>) => {
      const fields = recipe?.fields as RecipeFields;

      const id: string = recipe.sys.id;
      const title = fields?.title || "";
      const image = (fields?.photo?.fields?.file?.url || "") as string;
      const rating = 0;

      const recipeData: RecipesItemData = {
        id,
        title,
        image,
        rating,
      };

      return recipeData;
    }
  );

  return recipes;
};
