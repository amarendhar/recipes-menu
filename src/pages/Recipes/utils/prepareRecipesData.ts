import {
  Entry,
  EntryFields,
  EntryCollection,
  EntrySkeletonType,
} from "contentful";
import { Maybe, Recipe } from "types";

type RecipeFields = {
  title: EntryFields.Text;
  photo: EntryFields.AssetLink;
  tags: EntryFields.Array<EntryFields.EntryLink<EntrySkeletonType>>;
  description: EntryFields.Text;
  chef: EntryFields.EntryLink<EntrySkeletonType>;
};

export const prepareRecipes = (
  data: Maybe<EntryCollection<EntrySkeletonType, undefined, string>>
): Recipe[] => {
  if (!data?.items) {
    return [];
  }

  const recipes: Recipe[] = data.items.map(
    (recipe: Entry<EntrySkeletonType, undefined, string>) => {
      const fields = recipe?.fields as RecipeFields;

      const id: string = recipe.sys.id;
      const title = fields?.title || "";
      const image = (fields?.photo?.fields?.file?.url || "") as string;
      const tags = fields?.tags?.map(
        (tag) => (tag?.fields?.name as string) || ""
      );
      const description = fields?.description || "";
      const chefName = (fields?.chef?.fields?.name || "") as string;

      const recipeData: Recipe = {
        id,
        title,
        image,
        tags,
        description,
        chefName,
      };

      return recipeData;
    }
  );

  return recipes;
};
