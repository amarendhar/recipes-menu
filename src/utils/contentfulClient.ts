import { createClient, CreateClientParams } from "contentful";
import {
  CONTENTFUL_SPACE_ID,
  CONTENTFUL_ACCESS_TOKEN,
  CONTENTFUL_ENVIRONMENT_ID,
} from "globalConstants";

const clientParams: CreateClientParams = {
  space: CONTENTFUL_SPACE_ID,
  accessToken: CONTENTFUL_ACCESS_TOKEN,
};

if (CONTENTFUL_ENVIRONMENT_ID) {
  clientParams.environment = CONTENTFUL_ENVIRONMENT_ID;
}

export const client = createClient(clientParams);
