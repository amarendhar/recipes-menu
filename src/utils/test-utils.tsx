import { Route, Routes, BrowserRouter } from "react-router-dom";
import {
  render as rtlRender,
  renderHook as rtlRenderHook,
} from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import { defaultTheme } from "themes";
import { ContentfulProvider } from "api";
import App from "App";

const render = (
  ui: React.ReactElement,
  { route = "/" }: { route?: string | null } = {}
) => {
  window.history.pushState({}, "Test page", route);

  const Wrapper = ({ children }: { children: React.ReactNode }) =>
    route ? (
      <App />
    ) : (
      <ContentfulProvider>
        <ThemeProvider theme={defaultTheme}>
          <BrowserRouter>{children}</BrowserRouter>
        </ThemeProvider>
      </ContentfulProvider>
    );

  return {
    ...rtlRender(ui, {
      wrapper: Wrapper,
    }),
  };
};

const renderHook = (
  hook: any,
  { route = "/" }: { route?: string | null } = {}
) => {
  window.history.pushState({}, "Test page", route);

  const Wrapper = ({ children }: { children: any }) => (
    <ContentfulProvider>
      <BrowserRouter>
        <Routes>
          <Route path="/" element={children} />
          <Route path="/recipe/:id" element={children} />
        </Routes>
      </BrowserRouter>
    </ContentfulProvider>
  );

  return rtlRenderHook(() => hook(), { wrapper: Wrapper });
};

export * from "@testing-library/react";
export { render, renderHook };
