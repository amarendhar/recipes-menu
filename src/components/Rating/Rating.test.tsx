import { act, render, screen } from "utils/test-utils";
import { Rating } from ".";

describe("# Rating", () => {
  const initialProps = {
    value: 0,
    onChange: jest.fn(),
  };
  const renderComponent = () =>
    render(<Rating {...initialProps} />, { route: null });

  it("Should update rating, when rating is changed", async () => {
    const { container } = renderComponent();

    expect((screen.getByTestId("rating") as HTMLInputElement).value).toEqual(
      "0"
    );

    act(() => {
      screen.getByTestId("rating-3").click();
    });
    expect(initialProps.onChange).toHaveBeenCalledWith(3);
    expect((screen.getByTestId("rating") as HTMLInputElement).value).toEqual(
      "3"
    );

    expect(container).toMatchSnapshot();
  });

  it("Should re-render with new rating value, when new value-prop is passed", async () => {
    const newProps = {
      value: 3,
      onChange: jest.fn(),
    };

    const { rerender } = renderComponent();
    expect((screen.getByTestId("rating") as HTMLInputElement).value).toEqual(
      "0"
    );

    rerender(<Rating {...newProps} />);
    expect((screen.getByTestId("rating") as HTMLInputElement).value).toEqual(
      "3"
    );
  });
});
