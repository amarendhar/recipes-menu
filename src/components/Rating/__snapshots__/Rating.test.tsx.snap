// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`# Rating Should update rating, when rating is changed 1`] = `
<div>
  <div
    class="sc-jqUWZQ kSNpbf"
    data-testid="rating-container"
  >
    <button
      class="sc-kDDqET fgsgTg"
      data-testid="rating-1"
    >
      <svg
        height="24"
        style="color: rgb(250, 175, 0);"
        viewBox="0 0 24 24"
        width="24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
    </button>
    <button
      class="sc-kDDqET fgsgTg"
      data-testid="rating-2"
    >
      <svg
        height="24"
        style="color: rgb(250, 175, 0);"
        viewBox="0 0 24 24"
        width="24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
    </button>
    <button
      class="sc-kDDqET fgsgTg"
      data-testid="rating-3"
    >
      <svg
        height="24"
        style="color: rgb(250, 175, 0);"
        viewBox="0 0 24 24"
        width="24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
    </button>
    <button
      class="sc-kDDqET fgsgTg"
      data-testid="rating-4"
    >
      <svg
        height="24"
        style="color: rgba(0, 0, 0, 0.26);"
        viewBox="0 0 24 24"
        width="24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
    </button>
    <button
      class="sc-kDDqET fgsgTg"
      data-testid="rating-5"
    >
      <svg
        height="24"
        style="color: rgba(0, 0, 0, 0.26);"
        viewBox="0 0 24 24"
        width="24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
    </button>
    <input
      class="sc-iqcnOu ZBJeF"
      data-testid="rating"
      name="rating"
      type="radio"
      value="3"
    />
  </div>
</div>
`;
